### PR TITLE
Record username when flagging

### PR DESCRIPTION
**Pull request by abidlabs**
Fixes: #4134

Learned something very interesting when I was working on this. We were using `inspect.Parameter.annotation` in several places. This is a method that returns the type hint of a parameter. In Python 3.x, this returns a _class_ if the type hint is a class. [However, in the future, Python (4.x?), this will be replaced with a string](https://peps.python.org/pep-0563/). More importantly, if a user has:

```
from __future__ import annotations
```

in their code while running Python 3.x, this will also return a string (the future behavior). This was causing a subtle bug that took some time to figure out. Anyways, this is now fixed because we are using `typing.get_type_hints()` instead, which is the recommendation in the PEP.

### CODE REVIEW

File: gradio/helpers.py [hunk 1]
```diff
@@ -619,19 +620,18 @@ def special_args(
     progress_index = None
     event_data_index = None
     for i, param in enumerate(positional_args):
+        type_hint = type_hints.get(param.name)
```
Comment (abidlabs) [inline gradio/helpers.py:623]: See note in PR on why this was changed

File: gradio/helpers.py [hunk 2]
```diff
@@ -611,6 +611,7 @@ def special_args(
         updated inputs, progress index, event data index.
     """
     signature = inspect.signature(fn)
+    type_hints = utils.get_type_hints(fn)
```
Comment (abidlabs) [inline gradio/helpers.py:614]: See note in PR on why this was changed

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,21 +2,11 @@
 
 ## New Features:
 
-- Support for asynchronous iterators by [@freddyaboulton](https://github.com/freddyaboulton) in [PR 3821](https://github.com/gradio-app/gradio/pull/3821)
```
Comment (freddyaboulton) [inline CHANGELOG.md:5]: This line about the async generator got deleted by mistake

**Review by freddyaboulton — COMMENTED** [review]
Thanks for the fix @abidlabs !

This breaks the `HuggingFaceDatasetSaver` because it's flag method does not accept a username. Once that's in this should be good to merge.

**Comment by abidlabs** [comment]
Thanks @freddyaboulton for the review! I have made the fix so that usernames also work with the HF Savers. Thanks for catching that. Will merge once CI cooperates

**Pull request merged**